java/query/GraqlDefinable.java

```diff
@@ -30,43 +33,52 @@
-import static graql.lang.common.exception.ErrorMessage.MISSING_PATTERNS;
+import static graql.lang.common.exception.ErrorMessage.MISSING_DEFINABLES;
 import static java.util.stream.Collectors.joining;
 
 abstract class GraqlDefinable extends GraqlQuery {
 
     private final GraqlToken.Command keyword;
-    private final List<TypeVariable> variables;
+    private final List<TypeVariable> typeVariables = new ArrayList<>();
+    private final List<Rule> rules = new ArrayList<>();
     private final int hash;
 
-    GraqlDefinable(final GraqlToken.Command keyword, final List<TypeVariable> variables) {
+    GraqlDefinable(final GraqlToken.Command keyword, final List<Definable> definables) {
         assert keyword == DEFINE || keyword == UNDEFINE;
-        if (variables == null || variables.isEmpty()) throw GraqlException.of(MISSING_PATTERNS.message());
-        final LinkedList<TypeVariable> list = new LinkedList<>(variables);
-        while (!list.isEmpty()) {
-            final TypeVariable v = list.removeFirst();
+        if (definables == null || definables.isEmpty()) throw GraqlException.of(MISSING_DEFINABLES.message());
+        final LinkedList<TypeVariable> typeVars = new LinkedList<>();
+        for (Definable definable : definables) {
+            if (definable.isRule()) rules.add(definable.asRule());
+            if (definable.isTypeVariable()) typeVars.add(definable.asTypeVariable());
+        }
+        while (!typeVars.isEmpty()) {
+            final TypeVariable v = typeVars.removeFirst();
             if (!v.isLabelled()) throw GraqlException.of(ErrorMessage.INVALID_DEFINE_QUERY_VARIABLE.message());
-            else v.constraints().forEach(c -> list.addAll(c.variables()));
+            else v.constraints().forEach(c -> typeVars.addAll(c.variables()));
+            typeVariables.add(v);
         }
 
         this.keyword = keyword;
-        this.variables = variables;
-        this.hash = Objects.hash(this.keyword, this.variables);
+        this.hash = Objects.hash(this.keyword, this.typeVariables, this.rules);
     }
 
-    public final List<TypeVariable> variables() {
-        return variables;
+    public final List<TypeVariable> typeVariables() {
+        return typeVariables;
+    }
+    public final List<Rule> rules() {
+        return rules;
     }
 
     @Override
     public final String toString() {
         final StringBuilder query = new StringBuilder();
         query.append(keyword);
 
-        if (variables.size() > 1) query.append(NEW_LINE);
+        if (typeVariables.size() + rules.size() > 1) query.append(NEW_LINE);
         else query.append(GraqlToken.Char.SPACE);
 
-        query.append(variables().stream().map(TypeVariable::toString).collect(joining("" + SEMICOLON + NEW_LINE)));
+        query.append(typeVariables().stream().map(TypeVariable::toString).collect(joining("" + SEMICOLON + NEW_LINE)));
+        query.append(rules().stream().map(Rule::toString).collect(joining("" + SEMICOLON + NEW_LINE)));
```

Comment:
You've changed the order in which the user wrote the types and rules. If you just save the original list, and stream through that list in order, with an `if` switch to print different types, that would preserve the order, right?

Reply:
can just use `Definable` list and use `toString` on `Definable`!